Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+# Standard Repo API
+
+In order to build the .NET Core product from its constiuent repos, there needs to be a standardized set of actions
+each repo can perform. Not only will this make building the repos easier, it will make it easier to onboard
+new repos. If a new repo conforms to the standardized set of actions, source-build can easily start consuming it.
+
+This document describes the basic structure of this "Repo API", and the requirements imposed on a repo that needs
+to participate in source-build.
+
+## build.cmd/sh
+
+Each repo will have two files that implement the repo API:
+
+* build.cmd
+* build.sh
+
+The `.cmd` file will be called when running on Windows, while the `.sh` file will be called on all other operating systems.
+
+These scripts will contain the logic necessary in order to build, test, publish the repo's assets.
+
+source-build will invoke these scripts on each repo passing in:
+
+1. An action to perform
+1. A set of parameters
+
+Individual repos are free to implement more features in these build scripts than what is required by source-build. For
+example, a repo may want to provide a default action. These extra behaviors are not defined by source-build.
+
+### Actions
+
+To start, there are a number of actions source-build will need the repos to implement.  The actions are:
+
+1. `Restore` - Sync any external dependencies/tools necessary to build the repo.
+1. `Build` - Assemble the product from the current source contained in the repo.
+1. `Publish` - Put the assembled product in the specified location.
```

Comment:
Do we think we need an explicit publish verb or should it be handled by a property we pass to build which tells the build where to put stuff?

Reply:
I agree. This separation feels a bit artificial. Maybe `-p:DotNetOutputPath` passed to build?

Reply:
This verb is not about uploading artifacts, is it?

Reply:
Def dont need a publish verb. Instead we need a produces verb. Ex: what nuget packages did u prodsuce.

It is up to the orchestrator / source build to decide what to do with these items.

Reply:
>This verb is not about uploading artifacts, is it?

No this isn't about uploading the final artifacts as that will be handled by the root source-build or orchestrator.

Reply:
My initial thinking was that this was a "upload the artifacts" verb.  The reason being that today each repo puts their artifacts in differing spots:

* corefx uses the dotnet-core myget feed
* nuget uses the nuget-build myget feed
* asp.net uses different feeds, but none of the above
* etc

Also, core-setup and cli produce installers that go into Azure storage.  They also push .deb packages to our test feeds (`azure-apt-cat.cloudapp.net`).  Are we really going to have the orchestrator know how to publish all these different assets to all the different places?

Reply:
> Are we really going to have the orchestrator know how to publish all these different assets to all the different places?

The orchestrator is going to put good builds into known locations. The various teams can then afterwards move them to other locationt as they need. 

Def dont want to have the repos pushing the artifacts to all those places as a part of core build. That would mean builds end up getting blocked on items like inserting into VS. No one wants that.

Reply:
The orchestrator needs to know all the outputs of the overall build so it should be able to handle publishing the final build outputs. We don't want anything published until the entire build is successful either.

Reply:
They key in my mind is that each repo is self-declarative.  The anti-pattern is where we start using meta-data outside of the repo itself to compose the product overall.

Reply:
I'm going to take out both the `Restore` and `Publish` commands here.  They were really supposed to be examples, and not designing exactly how we were going to build either in orchestration or source-build mode.